src/wallet/walletdb.cpp

```diff
@@ -480,10 +502,24 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
             if (!ssValue.eof()) {
                 uint256 checksum;
                 ssValue >> checksum;
-                if ((checksum_valid = Hash(vchPrivKey) != checksum)) {
+                if (Hash(vchPrivKey) != checksum) {
                     strErr = "Error reading wallet database: Encrypted key corrupt";
                     return false;
                 }
+
+                // Get the signature and check it
+                if (!ssValue.eof()) {
+                    std::vector<unsigned char> sig;
+                    ssValue >> sig;
+                    // Our checksum is already the hash of the entire private key
+                    // so we can just use it as our message
+                    XOnlyPubKey xonly{vchPubKey};
+                    if (!xonly.VerifySchnorr(checksum, sig)) {
```

Comment:
If the deserialized signature doesn't have a size of 64 bytes, it would lead to a crash as the following assertion in `XOnlyPubKey::VerifySchnorr` will fail,
https://github.com/bitcoin/bitcoin/blob/b1a2021f78099c17360dc2179cbcb948059b5969/src/pubkey.cpp#L208
i.e. a signature size check before verifying should be added. As the same is needed below for `WALLETDESCRIPTORCKEY`, probably it would make sense to put the verification (including signature size check) into a helper function to deduplicate?

Reply:
Added that as well deduplicated.